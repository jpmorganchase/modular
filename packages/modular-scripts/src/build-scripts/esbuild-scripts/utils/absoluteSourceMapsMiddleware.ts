import * as path from 'path';
import * as fs from 'fs-extra';
import type { Request, Response } from 'express';

type SourceMap = { sources: string[] };

/* This middleware remedies this issue - https://github.com/evanw/esbuild/pull/1234,
   where paths in sources that are relative to outDir break react-error-overlay.
   This is to be used before static.
*/

export function createAbsoluteSourceMapMiddleware(outDir: string) {
  return function sourceMapMiddleware(req: Request, res: Response): void {
    const filePath = path.join(outDir, req.path);
    void fs.readJson(filePath).then((sourceMapObject: SourceMap) => {
      // Make all paths absolute
      sourceMapObject.sources = sourceMapObject.sources.map((sourcePath) =>
        path.join(outDir, sourcePath),
      );
      res.json(sourceMapObject);
    });
  };
}
