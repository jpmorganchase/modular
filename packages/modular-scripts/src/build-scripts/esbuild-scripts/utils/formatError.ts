import * as path from 'path';
import { Message } from 'esbuild';
import chalk from 'chalk';
import * as fs from 'fs-extra';
import { codeFrameColumns } from '@babel/code-frame';

export async function formatError(
  error: Message,
  baseDir: string = process.cwd(),
): Promise<string> {
  if (error.location?.file) {
    const pathToFile = path.join(baseDir, error.location?.file);
    const source = await fs.readFile(pathToFile, {
      encoding: 'utf-8',
    });
    return `${chalk.red('Error:')}[${error.location.file}] ${error.text}
  
${codeFrameColumns(
  source,
  { start: { line: error.location.line, column: error.location.column } },
  {
    highlightCode: true,
  },
)}
    `;
  } else {
    return `${chalk.red('Error:')} ${error.text}\n`;
  }
}
