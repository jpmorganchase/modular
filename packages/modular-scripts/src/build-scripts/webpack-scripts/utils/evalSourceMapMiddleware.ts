import Server from 'webpack-dev-server';
import webpack from 'webpack';
import type {
  NextFunction,
  Request,
  Response,
  Stats,
} from 'webpack-dev-server';

function getBase64SourceMap(source: webpack.sources.Source) {
  const base64 = Buffer.from(JSON.stringify(source.map()), 'utf8').toString(
    'base64',
  );
  return `data:application/json;charset=utf-8;base64,${base64}`;
}

function getSourceById(stats: Stats, id: string) {
  const module = Array.from(stats.compilation.modules).find(
    (m) => stats.compilation.chunkGraph.getModuleId(m) === id,
  );
  return module ? module.originalSource() : null;
}

/*
 * Middleware responsible for retrieving a generated source
 * Receives a webpack internal url: "webpack-internal:///<module-id>"
 * Returns a generated source: "<source-text><sourceMappingURL><sourceURL>"
 *
 * Based on EvalSourceMapDevToolModuleTemplatePlugin.js
 */
export default function createEvalSourceMapMiddleware(server: Server) {
  return function handleWebpackInternalMiddleware(
    req: Request,
    res: Response,
    next: NextFunction,
  ) {
    if (req.url.startsWith('/__get-internal-source')) {
      const fileName = req.query.fileName as string;
      const match = fileName.match(/webpack-internal:\/\/\/(.+)/);
      const id = match ? match[1] : undefined;
      // Server.stats is private so we get around it this way
      const stats = server['stats'] as Stats | undefined;
      if (!id || !stats) {
        next();
      } else {
        const source = getSourceById(stats, id);
        if (source) {
          const base64SourceMap = getBase64SourceMap(source);
          const sourceMapURL = `//# sourceMappingURL=${base64SourceMap}`;

          const sourceURL = `//# sourceURL=webpack-internal:///${module.id}`;
          res.end(
            `${source.source() as string}\n${sourceMapURL}\n${sourceURL}`,
          );
        }
      }
    } else {
      next();
    }
  };
}
