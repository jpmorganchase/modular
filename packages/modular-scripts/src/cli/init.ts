import { Command } from 'commander';
import { init } from '../';

interface Options {
  y: boolean;
}

const command = new Command('init')
  .description(
    `Initialize a new modular repository in the current working directory.`,
  )
  .option('-y', 'Equivalent of npm init -y when running modular init', false)
  .action((options: Options) => {
    return init(options.y);
  });

export default command;
