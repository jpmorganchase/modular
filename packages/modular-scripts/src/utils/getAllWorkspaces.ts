import execa from 'execa';
import memoize from './memoize';
import getModularRoot from './getModularRoot';
import stripAnsi from 'strip-ansi';

interface YarnWorkspace {
  location: string;
  workspaceDependencies: string[];
  mismatchedWorkspaceDependencies: string[];
}

type YarnWorkspaces = Record<string, YarnWorkspace>;

export function getYarnWorkspaceInfo(cwd: string): YarnWorkspaces {
  // strip out ANSI color codes and escape characters
  const strippedStd =
    stripAnsi(
      execa.sync('yarnpkg', ['--silent', 'workspaces', 'info'], {
        all: true,
        reject: false,
        cwd,
        cleanup: true,
      }).stdout,
    ) || '{}';

  return JSON.parse(strippedStd) as YarnWorkspaces;
}

function _getAllWorkspaces(): YarnWorkspaces {
  const modularRoot = getModularRoot();

  return getYarnWorkspaceInfo(modularRoot);
}

export const getAllWorkspaces = memoize(_getAllWorkspaces);
