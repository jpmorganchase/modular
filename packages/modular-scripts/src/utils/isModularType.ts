import { JSONSchemaForNPMPackageJsonFiles as PackageJson } from '@schemastore/package';
import * as path from 'path';
import * as fs from 'fs-extra';

export const packageTypes: PackageType[] = ['app', 'view', 'package'];

export const ModularTypes: ModularType[] = (
  packageTypes as ModularType[]
).concat(['root']);

export type PackageType = 'app' | 'view' | 'package';

export type ModularType = PackageType | 'root';

export type ModularPackageJson = PackageJson & {
  browserslist?: Record<string, string[]>;
  modular?: {
    type: ModularType;
  };
};

export function getModularType(dir: string): ModularType | undefined {
  const packageJsonPath = path.join(dir, 'package.json');
  if (fs.existsSync(packageJsonPath)) {
    const packageJson = fs.readJsonSync(packageJsonPath) as ModularPackageJson;
    return packageJson.modular?.type || 'package';
  }
}

export default function isModularType(dir: string, type: PackageType): boolean {
  const packageJsonPath = path.join(dir, 'package.json');
  if (fs.existsSync(packageJsonPath)) {
    const packageJson = fs.readJsonSync(packageJsonPath) as ModularPackageJson;
    return packageJson.modular?.type === type;
  }
  return false;
}

export function isValidModularType(dir: string): boolean {
  return ModularTypes.includes(getModularType(dir) as ModularType);
}

export function isValidModularRootPackageJson(dir: string): boolean {
  const packageJsonPath = path.join(dir, 'package.json');
  if (fs.existsSync(packageJsonPath)) {
    const packageJson = fs.readJsonSync(packageJsonPath) as ModularPackageJson;
    return (
      packageJson.modular?.type === 'root' &&
      !!packageJson.private &&
      !!packageJson?.workspaces?.includes('packages/**')
    );
  }
  return false;
}
