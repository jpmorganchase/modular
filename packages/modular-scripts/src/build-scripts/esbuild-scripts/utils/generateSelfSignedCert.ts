import fs from 'fs';
import path from 'path';
import { promisify } from 'util';
import selfsigned from 'selfsigned';
import rimraf from 'rimraf';
import * as logger from '../../../utils/logger';

const del = promisify(rimraf);

// This code is mostly duplicated from webpack-dev-server, with some typings added.
// To provide consistent https behaviour in esbuild mode, we use the same directories to store the generated cert.
// See: https://github.com/webpack/webpack-dev-server/blob/master/lib/Server.js

function findCacheDir() {
  const cwd = process.cwd();

  let dir: string | undefined = cwd;

  for (;;) {
    try {
      if (fs.statSync(path.join(dir, 'package.json')).isFile()) break;
    } catch (e) {}

    const parent = path.dirname(dir);

    if (dir === parent) {
      dir = undefined;
      break;
    }

    dir = parent;
  }

  if (!dir) {
    return path.resolve(cwd, '.cache/webpack-dev-server');
  } else if (process.versions.pnp === '1') {
    return path.resolve(dir, '.pnp/.cache/webpack-dev-server');
  } else if (process.versions.pnp === '3') {
    return path.resolve(dir, '.yarn/.cache/webpack-dev-server');
  }

  return path.resolve(dir, 'node_modules/.cache/webpack-dev-server');
}

export async function generateSelfSignedCert() {
  const certificateDir = findCacheDir();
  const certificatePath = path.join(certificateDir, 'server.pem');
  let certificateExists = fs.existsSync(certificatePath);

  if (certificateExists) {
    const certificateTtl = 1000 * 60 * 60 * 24;
    const certificateStat = fs.statSync(certificatePath);

    const now = new Date();

    // cert is more than 30 days old, kill it with fire
    if ((Number(now) - Number(certificateStat.ctime)) / certificateTtl > 30) {
      logger.log('SSL Certificate is more than 30 days old. Removing...');

      await del(certificatePath);

      certificateExists = false;
    }
  }

  if (!certificateExists) {
    logger.log('Generating SSL Certificate...');

    const attributes = [{ name: 'commonName', value: 'localhost' }];
    const pems = selfsigned.generate(attributes, {
      algorithm: 'sha256',
      days: 30,
      keySize: 2048,
      extensions: [
        {
          name: 'basicConstraints',
          cA: true,
        },
        {
          name: 'keyUsage',
          keyCertSign: true,
          digitalSignature: true,
          nonRepudiation: true,
          keyEncipherment: true,
          dataEncipherment: true,
        },
        {
          name: 'extKeyUsage',
          serverAuth: true,
          clientAuth: true,
          codeSigning: true,
          timeStamping: true,
        },
        {
          name: 'subjectAltName',
          altNames: [
            {
              // type 2 is DNS
              type: 2,
              value: 'localhost',
            },
            {
              type: 2,
              value: 'localhost.localdomain',
            },
            {
              type: 2,
              value: 'lvh.me',
            },
            {
              type: 2,
              value: '*.lvh.me',
            },
            {
              type: 2,
              value: '[::1]',
            },
            {
              // type 7 is IP
              type: 7,
              ip: '127.0.0.1',
            },
            {
              type: 7,
              ip: 'fe80::1',
            },
          ],
        },
      ],
    });

    fs.mkdirSync(certificateDir, { recursive: true });
    fs.writeFileSync(certificatePath, pems.private + pems.cert, {
      encoding: 'utf8',
    });
  }

  const fakeCert = fs.readFileSync(certificatePath);

  logger.log(`SSL certificate: ${certificatePath}`);

  return fakeCert;
}
