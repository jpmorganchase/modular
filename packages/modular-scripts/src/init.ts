import * as fs from 'fs-extra';
import * as path from 'path';
import execSync from './utils/execSync';
import { ModularPackageJson } from './utils/isModularType';

export async function initModularFolder(
  folder: string,
  initOverride: boolean,
): Promise<void> {
  const packageJsonPath = path.join(folder, 'package.json');

  if (!fs.existsSync(packageJsonPath)) {
    const args = ['init'];
    if (initOverride) {
      args.push('-y');
    }
    execSync('npm', args, {
      cwd: folder,
    });
  }

  const packageJson = (await fs.readJSON(
    packageJsonPath,
  )) as ModularPackageJson;

  let changed = false;
  if (!packageJson.modular) {
    packageJson.modular = {
      type: 'root',
    };
    changed = true;
  }

  if (!packageJson.workspaces) {
    packageJson.workspaces = ['packages/**'];
  }

  if (changed) {
    await fs.writeJSON(packageJsonPath, packageJson, {
      spaces: 2,
    });
  }

  await fs.mkdirp(path.join(folder, 'modular'));
  await fs.mkdirp(path.join(folder, 'packages'));

  console.log('Modular repository initialized!');
}

export default function init(initOverride = false): Promise<void> {
  return initModularFolder(process.cwd(), initOverride);
}
