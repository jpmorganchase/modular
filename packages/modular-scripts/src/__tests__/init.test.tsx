import { initModularFolder } from '../init';
import * as path from 'path';
import * as tmp from 'tmp';
import * as fs from 'fs-extra';
import { promisify } from 'util';
import { ModularPackageJson } from '../utils/isModularType';
import { getYarnWorkspaceInfo } from '../utils/getAllWorkspaces';

const mktempd = promisify(tmp.dir);

describe('Creating a new modular folder', () => {
  let folder: string;
  beforeEach(async () => {
    folder = await mktempd();
    await initModularFolder(folder, true);
  });

  afterEach(async () => {
    await fs.remove(folder);
  });

  it('should make a new repo with the right name and properties', async () => {
    const packageJson = (await fs.readJSON(
      path.join(folder, 'package.json'),
    )) as ModularPackageJson;

    packageJson.name = 'test-modular-app';

    expect(packageJson).toMatchSnapshot();
  });

  it('should copy over the modular root template files if they are not already present', async () => {
    const rootFiles = await fs.readdir(
      path.join(__dirname, '..', '..', 'types', 'root'),
    );
    rootFiles.forEach((name) => {
      expect(fs.existsSync(path.join(folder, name))).toBe(true);
    });
  });

  it('should have an empty yarn.lock', async () => {
    const lockfile = await fs.readFile(path.join(folder, 'yarn.lock'));

    expect(String(lockfile)).toMatchSnapshot();
  });

  it('should not have any workspace info', () => {
    const workspace = getYarnWorkspaceInfo(folder);
    expect(workspace).toEqual({});
  });
});
