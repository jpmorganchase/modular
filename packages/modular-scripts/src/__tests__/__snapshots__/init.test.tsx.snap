// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Creating a new modular folder should have an empty yarn.lock 1`] = `
"# THIS IS AN AUTOGENERATED FILE. DO NOT EDIT THIS FILE DIRECTLY.
# yarn lockfile v1


"
`;

exports[`Creating a new modular folder should make a new repo with the right name and properties 1`] = `
Object {
  "author": "",
  "description": "",
  "keywords": Array [],
  "license": "MIT",
  "main": "index.js",
  "modular": Object {
    "type": "root",
  },
  "name": "test-modular-app",
  "private": true,
  "scripts": Object {
    "test": "echo \\"Error: no test specified\\" && exit 1",
  },
  "version": "1.0.0",
  "workspaces": Array [
    "packages/**",
  ],
}
`;
